bam2m5: Covert bam file to Pacbio
====================================

Description
---------------

Pacbio .m5 alignment format is used by several assembly software like pbdagcon, sparc,
yet only blasr could output this format. sam/bam file is still the standard output
format for alignment. Here provides a converter from sam/bam format to m5.

Usage
-----------

Usage: python3 bam2m5.py <in.bam> <ref.fa> <score_scheme> <out.m5>

in.bam: input bam file
    - should be sorted by coordinate for effiency.
    - MD tag is needed. bwa results is with MD tag, for blasr, it could be generated by `samtools calmd`)
    - Note: as coordinate of read is used in m5 file, 
      if bam is generated by blasr, `-clipping (soft|hard)` parameter should be used. 

ref.fa: reference file.  
    <ref.fa>.fai should be pre generated by `samtools faidx`

score_scheme: 
    - scoring parameter used for alignment, 
      in format: match,mismatch,gap_open,gap_extend.
    - eg: -5,6,10,0 means score -5 for a match, score 6 for a mismatch, 
      10 for gap open and 0 for one base gap extend.

    - the default score schemes for these software:
        - blasr -sam : -5,6,10,0
        - blasr -m 5 : -5,6,0,5,
        - bwa mem -x pacbio: 1,-1,-1,-1

out.m5:
    output m5 file

example: python3 bam2m5.py align.sorted.bam ref.fa -5,6,0,5 align.sorted.m5

see test/run_test.sh for converting pabcio sam to m5

Dependency
------------

Python3_:
    this script is in python3, python2 support may be added later

BioUtil_ >= 0.2:
    handle bam file, fasta file reading
    use `pip3 install BioUtil` to install this pacakge.

Authors
----------

Yu XU, xuyu@genomics.cn


Lisense
-----------
These script is under GPL2 lisense

"""
